src/views/Search/TagFilter.tsx

```diff
@@ -1,31 +1,76 @@
-import { FunctionComponent } from "react";
-import { PackageTagConfig } from "../../api/config";
+import { FunctionComponent, useCallback, useMemo } from "react";
+import { PackageTagConfig, TagGroupConfig } from "../../api/config";
 import { useConfigValue } from "../../hooks/useConfigValue";
 import { CheckboxFilter } from "./CheckboxFilter";
+import { RadioFilter } from "./RadioFilter";
 import { useTags } from "./useSearchParam";
 import { useUpdateSearchParam } from "./useUpdateSearchParam";
 
+interface FilterGroup extends Partial<TagGroupConfig> {
+  id: string;
+  tags: PackageTagConfig[];
+}
 interface FilterGroups {
-  [group: string]: PackageTagConfig[];
+  [group: string]: FilterGroup;
 }
 
 export const TagFilter: FunctionComponent = () => {
-  const filterableTags =
-    useConfigValue("packageTags")?.filter((tag) => Boolean(tag.searchFilter)) ??
-    [];
-
-  const tagFilterGroups: FilterGroups = filterableTags.reduce(
-    (accum: FilterGroups, tag: PackageTagConfig): FilterGroups => {
-      const groupName = tag.searchFilter?.groupBy;
-      if (groupName) {
-        return {
-          ...accum,
-          [groupName]: [...(accum[groupName] ?? []), tag],
-        };
-      }
-      return accum;
-    },
-    {}
+  const packageTags = useConfigValue("packageTags");
+  const packageTagGroups = useConfigValue("packageTagGroups");
+
+  const filterableTags = useMemo(
+    () => packageTags?.filter((tag) => Boolean(tag.searchFilter)) ?? [],
+    [packageTags]
+  );
+
+  const tagGroupsMap = useMemo(() => {
+    const map = new Map<string, TagGroupConfig>();
+    packageTagGroups?.forEach((group) => {
+      map.set(group.id, group);
+    });
+
+    return map;
+  }, [packageTagGroups]);
+
+  const tagFilterGroups: FilterGroups = useMemo(
+    () =>
+      filterableTags.reduce(
+        (accum: FilterGroups, tag: PackageTagConfig): FilterGroups => {
```

Comment:
This useMemo function has a bit of complexity to it - can extract part of this into a utility function and unit test it perhaps?

Reply:
Good idea 👍